From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Wed, 10 Jan 2024 22:41:12 +0100
Subject: [PATCH] Modifiable Bukkit enums - Inject runtime versions - Material
 - Initialize data field

In the "Modifiable Bukkit enums - Remove hard-coded properties - Material - Data class" patch,
the `UnsafeValues.computeData` was added, returning the same value as is hard-coded as the `Material.data`
initialization. However, this initialization is no longer present in the runtime injected version of `Material`.

To use the `UnsafeValues.computeData` method to initialize the field, however, requires for `Bukkit.getUnsafe()`
to return a non-null value, which requires that the `Bukkit.server` singleton has been initialized.

Therefore, this patch initializes the `Material.data` field appropriately for all instances immediately after `Bukkit.server` is initialized.

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index b243db56756c67cd2c41d7768898d01539f9260a..8a093a5720b79c45f648504b780a286d546d28d3 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -106,6 +106,20 @@ public final class Bukkit {
         }
 
         Bukkit.server = server;
+        // Fiddle start - modifiable Bukkit enums - inject runtime versions - Material
+        // Initialize the Material data field values
+        if (server.getUnsafe() != null) { // Only if this is not a dummy server (used in unit tests)
+            try {
+                var dataField = Material.class.getDeclaredField("data");
+                dataField.setAccessible(true);
+                for (Material material : Material.values()) {
+                    dataField.set(material, material.computeData());
+                }
+            } catch (Exception e) {
+                throw new RuntimeException("Failed to initialize Material data field when setting singleton Server", e);
+            }
+        }
+        // Fiddle end - modifiable Bukkit enums - inject runtime versions - Material
         // Paper start - add git information
         server.getLogger().info(getVersionMessage());
     }
